spark/src/test/scala-spark-master/org/apache/spark/sql/delta/DeltaVariantSuite.scala

```diff
@@ -126,4 +137,417 @@ class DeltaVariantSuite
       )
     }
   }
+
+  test("streaming variant delta table") {
+    withTempDir { dir =>
+      val path = dir.getAbsolutePath
+      spark.range(100)
+        .selectExpr("parse_json(cast(id as string)) v")
+        .write
+        .format("delta")
+        .mode("overwrite")
+        .save(path)
+
+      val streamingDf = spark.readStream
+        .format("delta")
+        .load(path)
+        .selectExpr("v::int as extractedVal")
+
+      val q = streamingDf.writeStream
+        .format("memory")
+        .queryName("test_table")
+        .start()
+
+      q.processAllAvailable()
+      q.stop()
+
+      val actual = spark.sql("select extractedVal from test_table")
+      val expected = spark.sql("select id from range(100)")
+      checkAnswer(actual, expected.collect())
+    }
+  }
+
+  test("variant works with schema evolution") {
+    withTempDir { dir =>
+      val path = dir.getAbsolutePath
+      spark.range(0, 100, 1, 1)
+        .selectExpr("id", "parse_json(cast(id as string)) v")
+        .write
+        .format("delta")
+        .mode("overwrite")
+        .save(path)
+
+      spark.range(100, 200, 1, 1)
+        .selectExpr(
+          "id",
+          "parse_json(cast(id as string)) v",
+          "parse_json(cast(id as string)) v_two"
+        )
+        .write
+        .format("delta")
+        .mode("append")
+        .option("mergeSchema", "true")
+        .save(path)
+
+      val expected = spark.range(0, 200, 1, 1).selectExpr(
+        "id",
+        "parse_json(cast(id as string)) v",
+        "case when id >= 100 then parse_json(cast(id as string)) else null end v_two"
+      )
+
+      val read = spark.read.format("delta").load(path)
+      checkAnswer(read, expected.collect())
+    }
+  }
+
+  test("variant cannot be used as a clustering column") {
+    withTable("tbl") {
+      val e = intercept[DeltaAnalysisException] {
+        sql("CREATE TABLE tbl(v variant) USING DELTA CLUSTER BY (v)")
+      }
+      checkError(
+        e,
+        "DELTA_CLUSTERING_COLUMN_MISSING_STATS",
```

Comment:
This fails because of missing stats but tests in `StatsCollectionSuite` show that we can compute stats for variant columns.
Is this test just not doing some pre-requisite steps to be able to use variant as clustering key or do we actually reject this?

Reply:
IIUC, clustering requires the type to be comparable. Currently variants are not comparable so the min/max stats are missing. null count stats, however, are present.